Automattic/woocommerce-payments
# Client Side fetching and caching of currency rates

Fixes 793-gh-Automattic/woocommerce-payments-server

#### Changes proposed in this Pull Request

- Adds a way of fetching currency rates from the server.
- Adds a mechanism to cache currency rates
- Replaces the mock currency with a call to the server/cache to get live rates.

#### Notes

- We had some discussions regarding how to architecture this so that we can pass an instance of the API Client into the `Multi_Currency` singleton. In the end, I went with the simple solution of having a static function which returns the API client, and using this in the `instance()` function for `Multi_Currency`. I don't think this is the best/most elegant solution, but I thought that as we currently have an open discussion regarding moving to PSR-4, with @marcinbot already opening a [great PR](https://github.com/Automattic/woocommerce-payments/pull/2061/files) for this, it would be better to tackle this issue either as a new issue, or as part of the PSR-4 PR, depending on the outcome of the discussions to move in that direction.

Some other issues I discovered while testing which aren't strictly related to this PR but were surfaced by it which I'll open new issues for:
- Some flags are not displaying (e.g. West African Francs) - we'll need to decide how to handle this. (see screenshot below)
- I noticed we have BitCoin in the list of currencies because it is returned by `get_woocommerce_currencies()`. I also noticed that weirdly, BitCoin has the flag of Bhutan in the list. I think we should just exclude BitCoin from our currency list, unless we do in fact want to support payments in it?
- Some character symbols aren't rendering properly in the Add Currencies modal, for example, see how Vietnamese đồng renders in the screenshot below.
- A server error "Error retrieving USD:VEF" is logged with every request. We do have server logic to check if currencies are supported by openexchange before making the request, so I think this will be a simple update of the supported currencies list, since VEF (Venezuelan Bolivar) must be on the list, but not actually supported by OpenExchange any more. 
- The `update` time is returned with the list of currencies, so we can now use this to populate the "Last Updated" time which displays on the Edit Currency page.

<img width="555" alt="Screenshot 2021-06-10 at 11 35 04" src="https://user-images.githubusercontent.com/11770181/121511141-3ee53c80-c9e0-11eb-853b-40c7530964b9.png">

#### Testing instructions

*Test 1*

- Checkout this branch, in the dev tools, disable API request redirection so that you're making requests to either the live API or your sandbox (depending on your `/etc/hosts` file configuration).
- If using sandbox, update it by running `svn up .` from the `public_html` folder to make sure you have the latest version of the server code.
- Go to the main Multi Currency page in the settings. Click on the `Add Currencies` modal. You should see a large list of currencies (~162). Test enabling a few of these currencies and verify the rates look sensible (note - if we can't fetch a rate for a particular currency, the automatic rate will be set to 0 meaning the user will need to set a manual rate if they wish to use that currency).
- Check the value of the `wcpay_multi_currency_cached_currencies` option either in the DB or by running `get_option('wcpay_multi_currency_cached_currencies');`. Verify it contains the array of currencies, plus the `updated` time and the `expires` time. The `expires` time should be 6 hours after the `updated` time. 

*Test 2*

- Delete the `wcpay_multi_currency_cached_currencies` option.
- Add the following line of code to line 244 of the `class-multi-currency.php` file:
```php
throw new API_Exception( 'error', 'error', 500);
```
- Refresh the multi currency page. Only the store currency should appear in enabled currencies, because we failed to fetch the currencies from the server.
- Check the value of the `wcpay_multi_currency_cached_currencies` option. The `currencies` value should be "error", and the `expires` time should be 1 minute after the `updated` time.
- Wait a minute, then remove the line you added to line 244 of `class-multi-currency.php`. Refresh the page. The currencies should be correctly fetched from the server.

- Do a quick smoke test of the multi-currency pages to make sure things are working as expected.
-------------------
